### PR TITLE
Add skin editor saving / loading support

### PR DESCRIPTION
**Pull request by peppy**
This is intended to be a first pass to get the skin editor into a state where users can consume it in a release. As with other large PRs, I've kept each commit to be quite self contained (especially towards the end - some of the earlier commits are often reverting changes, so I'd recommend only drilling into individual commits if you have a point of concern).

- [x] Depends on #12754 
- [x] Depends on #12753 
- [x] Depends on #12752 

https://user-images.githubusercontent.com/191335/117798309-faa13800-b28b-11eb-942d-8cad35318632.mp4

A few remaining tasks (mostly UX) which will be done in follow-up PRs to keep the size down:

- Ensure elements stay on screen
- Make blueprints visible even when not selected
- Pause gameplay or something to make it easier to edit without losing changes
- Update all skinnable test scenes and remove `SkinnableComboCounter` etc.
- Show anchors and origins on blueprints, and make blueprints always visible (so the user knows which components they can modify)
- Investigate whether `IStorageResourceProvider.Files` can resolve using a local path
  - Currently it looks like we are using a local path resolution (ie. from a beatmap or skin) where we probably don't need access to the global store anyway. Made apparent when looking at `LegacySkin`'s constructor, which actually provides a secondary store with local files which is used in most cases anyway...

### CODE REVIEW

File: osu.Game/Extensions/DrawableExtensions.cs [hunk 1]
```diff
@@ -43,5 +45,23 @@ public static ScheduledDelegate BeginKeyRepeat(this IKeyBindingHandler handler,
...
+            component.Scale = info.Scale;
+            component.Anchor = info.Anchor;
+            component.Origin = info.Origin;
+
+            if (component is Container container)
+            {
+                foreach (var child in info.Children)
+                    container.Add(child.CreateInstance());
+            }
+        }
```
Comment (peppy) [inline osu.Game/Extensions/DrawableExtensions.cs:65]: I couldn't decide whether these should be `DrawableExtensions` or `SkinnableComponentExtensions`. Open to moving them. The hope is that eventually `IDrawable` can be expanded to cover all the requirements of this PR so we don't need to ever cast to `Drawable`, though.

File: osu.Game/Screens/Edit/Compose/Components/BlueprintContainer.cs [hunk 2]
```diff
@@ -24,6 +26,7 @@ namespace osu.Game.Screens.Edit.Compose.Components
     /// Includes selection and manipulation support via a <see cref="Components.SelectionHandler{T}"/>.
     /// </summary>
     public abstract class BlueprintContainer<T> : CompositeDrawable, IKeyBindingHandler<PlatformAction>
+        where T : class
```
Comment (peppy) [inline osu.Game/Screens/Edit/Compose/Components/BlueprintContainer.cs:29]: Done to allow the reference comparisons below.

File: osu.Game/Screens/Play/HUD/SkinnableInfo.cs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        {
+            Type = component.GetType();
+
+            Position = component.Position;
+            Rotation = component.Rotation;
+            Scale = component.Scale;
+            Anchor = component.Anchor;
+            Origin = component.Origin;
+
+            if (component is Container container)
```
Comment (peppy) [inline osu.Game/Screens/Play/HUD/SkinnableInfo.cs:50]: Might be safer as

```csharp
            if (component is IReadOnlyList<Drawable> container)
```

Would be nice to somehow handle `Container<T>` types which use custom drawables, but not sure if that's possible.

File: osu.Game/Skinning/DefaultSkin.cs [hunk 4]
```diff
@@ -23,17 +27,77 @@ public DefaultSkin(IStorageResourceProvider resources)
...
+            if (base.GetDrawableComponent(component) is Drawable c)
+                return c;
+
+            switch (component)
+            {
+                case SkinnableTargetComponent target:
+                    switch (target.Target)
+                    {
+                        case SkinnableTarget.MainHUDComponents:
+                            var skinnableTargetWrapper = new SkinnableTargetWrapper(container =>
```
Comment (peppy) [inline osu.Game/Skinning/DefaultSkin.cs:50]: Importantly, this is never actually reached if the user has already customised the skin (even for the `DefaultSkin`). The user's layout will be returned in the `base.` call above.

**Comment by peppy** [comment]
I've finished a self-review pass of this, so it should be in a stable state for review now.

**Comment by ekrctb** [comment]
Not about this PR's change, but I think `SkinEditor` test scene looks better if autoplay is enabled, as the empty HP bar is difficult to see in the default skin.

**Comment by bdach** [comment]
@ekrctb #12672 will alleviate that somewhat.

**Comment by smoogipoo** [comment]
I can't repro so I don't think it's related to this branch, but I encountered a super dangerous crash pressing my left mouse button on the score counter:
```
[runtime] 2021-05-13 06:18:49 [error]: at System.Diagnostics.TraceInternal.Fail(String message)
[runtime] 2021-05-13 06:18:49 [error]: at osu.Framework.Input.Bindings.KeyBindingContainer`1.handleNewReleased(InputState state, InputKey releasedKey)
[runtime] 2021-05-13 06:18:49 [error]: at osu.Framework.Input.Bindings.KeyBindingContainer`1.Handle(UIEvent e)
[runtime] 2021-05-13 06:18:49 [error]: at osu.Framework.Graphics.Drawable.OnMouseUp(MouseUpEvent e)
[runtime] 2021-05-13 06:18:49 [error]: at osu.Framework.Graphics.Drawable.TriggerEvent(UIEvent e)
[runtime] 2021-05-13 06:18:49 [error]: at osu.Framework.Input.ButtonEventManager`1.<>c__DisplayClass14_0.<PropagateButtonEvent>b__0(Drawable target)
[runtime] 2021-05-13 06:18:49 [error]: at System.Linq.Enumerable.TryGetFirst[TSource](IEnumerable`1 source, Func`2 predicate, Boolean& found)
[runtime] 2021-05-13 06:18:49 [error]: at System.Linq.Enumerable.FirstOrDefault[TSource](IEnumerable`1 source, Func`2 predicate)
[runtime] 2021-05-13 06:18:49 [error]: at osu.Framework.Input.ButtonEventManager`1.PropagateButtonEvent(IEnumerable`1 drawables, UIEvent e)
[runtime] 2021-05-13 06:18:49 [error]: at osu.Framework.Input.MouseButtonEventManager.HandleButtonUp(InputState state, List`1 targets)
[runtime] 2021-05-13 06:18:49 [error]: at osu.Framework.Input.ButtonEventManager`1.handleButtonUp(InputState state)
[runtime] 2021-05-13 06:18:49 [error]: at osu.Framework.Input.ButtonEventManager`1.HandleButtonStateChange(InputState state, ButtonStateChangeKind kind)
[runtime] 2021-05-13 06:18:49 [error]: at osu.Framework.Input.InputManager.HandleMouseButtonStateChange(ButtonStateChangeEvent`1 e)
[runtime] 2021-05-13 06:18:49 [error]: at osu.Framework.Input.UserInputManager.HandleInputStateChange(InputStateChangeEvent inputStateChange)
[runtime] 2021-05-13 06:18:49 [error]: at osu.Framework.Input.StateChanges.ButtonInput`1.Apply(InputState state, IInputStateChangeHandler handler)
[runtime] 2021-05-13 06:18:49 [error]: at osu.Framework.Input.InputManager.Update()
[runtime] 2021-05-13 06:18:49 [error]: at osu.Framework.Input.PassThroughInputManager.Update()
[runtime] 2021-05-13 06:18:49 [error]: at osu.Framework.Graphics.Drawable.UpdateSubTree()
[runtime] 2021-05-13 06:18:49 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
[runtime] 2021-05-13 06:18:49 [error]: at osu.Framework.Platform.GameHost.UpdateFrame()
[runtime] 2021-05-13 06:18:49 [error]: at osu.Framework.Threading.GameThread.ProcessFrame()
```

**Comment by peppy** [comment]
I've managed to repro this one on this branch, but I don't think it's something I'm doing wrong locally to the editor. Probably a larger issue with the `KeyBindingContainer` or `PassThroughInputManager` handling (there's a few similar issues on o!f side, although the stack doesn't match 1:1).

**Comment by smoogipoo** [comment]
Here's another one, also probably not related to this branch?
```
[runtime] 2021-05-13 06:21:51 [error]: System.InvalidOperationException: Sequence contains more than one matching element
[runtime] 2021-05-13 06:21:51 [error]: at System.Linq.ThrowHelper.ThrowMoreThanOneMatchException()
[runtime] 2021-05-13 06:21:51 [error]: at System.Linq.Enumerable.SingleOrDefault[TSource](IEnumerable`1 source, Func`2 predicate)
[runtime] 2021-05-13 06:21:51 [error]: at osu.Game.Screens.Edit.Compose.Components.SelectionBoxDragHandleContainer.updateRotationHandlesVisibility() in /home/smgi/Repos/osu/osu.Game/Screens/Edit/Compose/Components/SelectionBoxDragHandleContainer.cs:line 88
[runtime] 2021-05-13 06:21:51 [error]: at osu.Game.Screens.Edit.Compose.Components.SelectionBoxDragHandle.OnHover(HoverEvent e) in /home/smgi/Repos/osu/osu.Game/Screens/Edit/Compose/Components/SelectionBoxDragHandle.cs:line 43
[runtime] 2021-05-13 06:21:51 [error]: at osu.Framework.Input.InputManager.updateHoverEvents(InputState state)
[runtime] 2021-05-13 06:21:51 [error]: at osu.Framework.Input.InputManager.HandleMousePositionChange(MousePositionChangeEvent e)
[runtime] 2021-05-13 06:21:51 [error]: at osu.Framework.Input.StateChanges.MousePositionAbsoluteInput.Apply(InputState state, IInputStateChangeHandler handler)
[runtime] 2021-05-13 06:21:51 [error]: at osu.Framework.Input.PassThroughInputManager.Handle(UIEvent e)
[runtime] 2021-05-13 06:21:51 [error]: at osu.Framework.Input.InputManager.PropagateBlockableEvent(SlimReadOnlyListWrapper`1 drawables, UIEvent e)
[runtime] 2021-05-13 06:21:51 [error]: at osu.Framework.Input.InputManager.handleMouseMove(InputState state, Vector2 lastPosition)
[runtime] 2021-05-13 06:21:51 [error]: at osu.Framework.Input.InputManager.HandleMousePositionChange(MousePositionChangeEvent e)
[runtime] 2021-05-13 06:21:51 [error]: at osu.Framework.Input.UserInputManager.HandleInputStateChange(InputStateChangeEvent inputStateChange)
[runtime] 2021-05-13 06:21:51 [error]: at osu.Framework.Input.StateChanges.MousePositionAbsoluteInput.Apply(InputState state, IInputStateChangeHandler handler)
[runtime] 2021-05-13 06:21:51 [error]: at osu.Framework.Input.InputManager.Update()
[runtime] 2021-05-13 06:21:51 [error]: at osu.Framework.Input.PassThroughInputManager.Update()
[runtime] 2021-05-13 06:21:51 [error]: at osu.Framework.Graphics.Drawable.UpdateSubTree()
[runtime] 2021-05-13 06:21:51 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
[runtime] 2021-05-13 06:21:51 [error]: at osu.Framework.Platform.GameHost.UpdateFrame()
[runtime] 2021-05-13 06:21:51 [error]: at osu.Framework.Threading.GameThread.ProcessFrame()
```
Happens when randomly dragging (quick click + drag + release sequences) on elements. Not sure exactly.

**Comment by peppy** [comment]
That one's fixed via proposed #12768

**Comment by ekrctb** [comment]
The biggest concern is the serialization of `Type` and usage of `ActivateInstance`.
This is fragile (changing type name or removing default constructor will break the skin without notice) and potentially dangerous (class not intended for skinning is instantiated).
I'm not sure about implementation, but I think `ISkin.GetDrawableComponent` should be used for creating drawable. `ISkinComponent` is serialized.
Custom properties (position etc.) are then applied to the resulting drawable.

**Comment by peppy** [comment]
Thing is, we do actually intend to allow skin creators to make their own drawables as a step beyond this PR. That will not come for a while, but *will* require some level of `Activator` usage.

I agree this is quite dangerous as it stands, but not sure limiting to a known subset is worthwhile at this point in time. I think we may be better off adding sanity via interface types (ie. ensure the constructed instances are `ISkinnableComponent`, and adding a property to that interface which specifies where the component is allowed to exist - this is actually something I plan to add in the near future to allow multiple areas of the game to be skinned, and components to correctly be populated into the toolbox).

As for class renames, let's just avoid doing that. As in, make sure we have finalised class names/namespaces before exposing them as user customisable.

**Comment by bdach** [comment]
Could also sprinkle `[UsedImplicitly]` on all classes that are instantiated via activator on our side, but I do agree it is a bit brittle still. That is probably the price to be paid for extensibility at this scope either way.

**Comment by peppy** [comment]
Not sure that will add much clarity. I think the interface specification is likely enough..

**Comment by ekrctb** [comment]
Another "open" (allowing external assembly to add custom `Drawable`) approach is have a registry of types suited for the skin customization. Probably attribute is added for each such class. A stable name can be specified as well. Like:

```csharp
[SkinComponent(stableName: "LegacyComboCounter")]
class LegacyComboCounter {
...
```

**Comment by peppy** [comment]
Not sure if attribute or interface work better, but sure, we can decide which of the two we prefer after we have more usage of the system.

File: osu.Game/Skinning/DefaultSkin.cs (hunk 4 above)
Comment (frenzibyte) [reply osu.Game/Skinning/DefaultSkin.cs:50]: Yeah I was going to touch on this, I'm not sure if you intentionally avoided it, but since the order of the drawable component retrieval has been flipped around to prefer customized components over default, it all feels like the skinnable target content logic isn't really suitable for this.

Would like to hear your opinion in moving all this into new separate methods specific for it (and obsolete/remove `GetDrawableComponent` if that's the way forward after the transition completes, which I presume to happen):
```csharp
class Skin
{
    // returns customized skin components for given target, 
    // or GetDefaultSkinComponents(target) if there isn't.
    public SkinnableTargetWrapper GetSkinComponents(SkinnableTarget target);
    
    // returns default skin components for given target.
    protected abstract SkinnableTargetWrapper GetDefaultSkinComponents(SkinnableTarget target);
}
```

(not sure about using `SkinnableTargetWrapper` as return type outright though, maybe `Container` instead?)

Here's a branch https://github.com/ppy/osu/compare/master...frenzibyte:refactor-skin-components-retrieval if it makes it any easier to do it.

Comment (peppy) [reply osu.Game/Skinning/DefaultSkin.cs:50]: I'm not sure what you mean by "doesn't feel suited"? Is there some expectation that the base lookup should be last? What's the rationale here?

File: osu.Game/Skinning/LegacyAccuracyCounter.cs [hunk 5]
```diff
@@ -12,37 +12,25 @@ namespace osu.Game.Skinning
 {
     public class LegacyAccuracyCounter : GameplayAccuracyCounter, ISkinnableComponent
     {
-        private readonly ISkin skin;
+        [Resolved]
+        private ISkinSource skin { get; set; }
```
Comment (smoogipoo) [inline osu.Game/Skinning/LegacyAccuracyCounter.cs:16]: What's this used for?

File: osu.Game/Skinning/ISkinSerialisable.cs [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+
+using osu.Framework.Graphics;
+
+namespace osu.Game.Skinning
+{
+    /// <summary>
+    /// Denotes a drawable component which should be serialised as part of a skin.
+    /// Use <see cref="ISkinnableComponent"/> for components which should be mutable by the user / editor.
+    /// </summary>
+    public interface ISkinSerialisable : IDrawable
```
Comment (smoogipoo) [inline osu.Game/Skinning/ISkinSerialisable.cs:12]: I don't really understand this distinction between `ISkinSerialisable` and `ISkinnableComponent`, and the xmldoc doesn't explain it. Why does `ISkinSerialisable` need to exist, and `ISkinnableComponent` can't be used? They're both empty interfaces.
More pragmatically, a drawable which "can be adjusted via skinning specifications" (`ISkinnableComponent`) must be serialised anyway, no?

File: osu.Game/Skinning/SkinManager.cs [hunk 7]
```diff
@@ -152,6 +155,52 @@ private void populateMetadata(SkinInfo item, Skin instance)
...
+
+                    var oldFile = skin.SkinInfo.Files.FirstOrDefault(f => f.Filename == filename);
+
+                    if (oldFile != null)
+                        ReplaceFile(skin.SkinInfo, oldFile, streamContent, oldFile.Filename);
+                    else
+                        AddFile(skin.SkinInfo, streamContent, filename);
+
+                    Logger.Log($"Saving out {filename} with {json.Length} bytes");
+                    Logger.Log(json);
```
Comment (smoogipoo) [inline osu.Game/Skinning/SkinManager.cs:199]: Probably remove?

File: osu.Game/Skinning/SkinnableElementTargetContainer.cs [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+        /// <param name="component">The component to add.</param>
+        /// <exception cref="NotSupportedException">Thrown when attempting to add an element to a target which is not supported by the current skin.</exception>
+        /// <exception cref="ArgumentException">Thrown if the provided instance is not a <see cref="Drawable"/>.</exception>
+        public void Add(ISkinnableComponent component)
+        {
+            if (content == null)
+                throw new NotSupportedException("Attempting to add a new component to a target container which is not supported by the current skin.");
+
+            if (!(component is Drawable drawable))
+                throw new ArgumentException("Provided argument must be of type {nameof(ISkinnableComponent)}.", nameof(drawable));
```
Comment (smoogipoo) [inline osu.Game/Skinning/SkinnableElementTargetContainer.cs:61]: Check does not match exception text.

File: osu.Game/Skinning/SkinnableTargetComponent.cs [hunk 9]
```diff
@@ -0,0 +1,17 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+namespace osu.Game.Skinning
+{
+    public class SkinnableTargetComponent : ISkinComponent
```
Comment (smoogipoo) [inline osu.Game/Skinning/SkinnableTargetComponent.cs:6]:
```
ISkinnableComponent
ISkinComponent
ISkinSerialisable
ISkinnableTarget

SkinnableTargetComponent : ISkinComponent (doesn't inherit either ISkinnableComponent or ISkinnableTarget??)
SkinnableTarget (doesn't inherit ISkinnableTarget)
```
This is all very hard to follow because naming is a mess.

I'm not sure on the name of `SkinnableTargetWrapper` and `SkinnableElementTargetWrapper` either. Like there's one-too-many `Target`s here.

File: osu.Game/Skinning/DefaultSkin.cs (hunk 4 above)
Comment (frenzibyte) [reply osu.Game/Skinning/DefaultSkin.cs:50]: `GetDrawableComponent()` is usually known to call its base method last when no matching component is found, so it looks unusual to suddenly make it called first, though I wouldn't really call this a valid reason. 

But, I think defining specific methods for this would be better to avoid any confusion/misunderstanding on the `SkinnableTargetComponent` case handling in skin implementations, for the fact they return the default components and will be overridden by customized components in base method.

Either way, at this point, this may need more opinions, as I'm not really sure which way is better anymore.

Comment (peppy) [reply osu.Game/Skinning/DefaultSkin.cs:50]: If the change you propose is to be made, it should be on the existing method, not a new one. There's no reason to split it out into its own method unless we decide to do the same for the other lookup types, or decide we don't want the type-based lookup in the first place.

File: osu.Game/Skinning/SkinnableTargetWrapper.cs [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+using Newtonsoft.Json;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+
+namespace osu.Game.Skinning
+{
+    /// <summary>
+    /// A container which is serialised and can encapsulate multiple skinnable elements into a single return type (for consumption via <see cref="ISkin.GetDrawableComponent"/>.
+    /// Will also optionally apply default cross-element layout dependencies when initialised from a non-deserialised source.
+    /// </summary>
```
Comment (smoogipoo) [inline osu.Game/Skinning/SkinnableTargetWrapper.cs:14]: Rather than explaining that it's serialised, perhaps this should explain that this contains the elements of a `SkinnableTarget`? As far as I understand, that's what this is doing.

File: osu.Game/Skinning/SkinnableTargetWrapper.cs [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+    [Serializable]
+    public class SkinnableTargetWrapper : Container, ISkinSerialisable
+    {
+        private readonly Action<Container> applyDefaults;
+
+        /// <summary>
+        /// Construct a wrapper with defaults that should be applied once.
+        /// </summary>
+        /// <param name="applyDefaults">A function with default to apply after the initial layout (ie. consuming autosize)</param>
```
Comment (smoogipoo) [inline osu.Game/Skinning/SkinnableTargetWrapper.cs:23]: What does "consuming" autosize mean?

**Review by smoogipoo — CHANGES_REQUESTED** [review]

File: osu.Game/Skinning/SkinnableTargetWrapper.cs (hunk 11 above)
Comment (peppy) [reply osu.Game/Skinning/SkinnableTargetWrapper.cs:23]: For cases like score counters, they require a value to be filled via the bindable flow to get a size (via `SpriteText`'s `AutoSize`). Maybe badly worded here.

File: osu.Game/Skinning/SkinnableTargetWrapper.cs (hunk 10 above)
Comment (peppy) [reply osu.Game/Skinning/SkinnableTargetWrapper.cs:14]: Sure, open for rewording. Provide a suggestion to reduce back-and-forth if possible.

File: osu.Game/Skinning/SkinnableTargetWrapper.cs (hunk 11 above)
Comment (smoogipoo) [reply osu.Game/Skinning/SkinnableTargetWrapper.cs:23]: Maybe just make it say "a function to apply the initial layout"? The line is pretty badly worded in general :P

File: osu.Game/Skinning/SkinnableTargetComponent.cs (hunk 9 above)
Comment (peppy) [reply osu.Game/Skinning/SkinnableTargetComponent.cs:6]: The last two are definitely pain points. I was following existing naming but agree they don't make things any clearer. The first is a "lookup" type. Should we renaming that whole hierarchy to SkinLookup instead of SkinComponent?

We could also choose a synonym for `Target` like `Area` or something.

File: osu.Game/Skinning/ISkinSerialisable.cs (hunk 6 above)
Comment (peppy) [reply osu.Game/Skinning/ISkinSerialisable.cs:12]: `ISkinSerialisable` is serialised but ignored by the editor.

Comment (smoogipoo) [reply osu.Game/Skinning/ISkinSerialisable.cs:12]: Can't you add an `IsEditable` property? Using interfaces as a tagging mechanism like this always leads to confusion.

Comment (peppy) [reply osu.Game/Skinning/ISkinSerialisable.cs:12]: Yep, almost went in that direction, but it mean that this will need to be added to each skin component. If you prefer that then I'll make the change.

For reference:

20210513 164543 (JetBrains Rider)

Comment (smoogipoo) [reply osu.Game/Skinning/ISkinSerialisable.cs:12]: You can implement a default member for the interface:
```csharp
public interface ISkinnableComponent
{
    bool IsEditable => true;
}
```
Because overrides to false are only done in two particular places, ever.

File: osu.Game/Skinning/SkinnableTargetComponent.cs (hunk 9 above)
Comment (frenzibyte) [reply osu.Game/Skinning/SkinnableTargetComponent.cs:6]: Since `SkinnableTargetWrapper` is for components of a given target, how about `SkinnableComponentsContainer` instead?

Comment (smoogipoo) [reply osu.Game/Skinning/SkinnableTargetComponent.cs:6]: > The first is a "lookup" type

Funny, because it's actually the _second_ that is the lookup type :stuck_out_tongue:

But I mean, `ISkinComponent` is more what I'd expect for a lookup type than I'd expect `ISkinnableComponent` to be a drawable type. Almost feels like `ISkinnableComponent` should be called `ISkinnable`.

File: osu.Game/Skinning/Skin.cs [hunk 12]
```diff
@@ -27,9 +35,65 @@ public abstract class Skin : IDisposable, ISkin
...
+        }
+
+        /// <summary>
+        /// Update serialised information for the provided target.
+        /// </summary>
+        /// <param name="targetContainer">The target container to serialise to this skin.</param>
+        public void UpdateDrawableTarget(SkinnableElementTargetContainer targetContainer)
+        {
+            DrawableComponentInfo[targetContainer.Target] = targetContainer.CreateSkinnableInfo().ToArray();
+        }
```
Comment (frenzibyte) [inline osu.Game/Skinning/Skin.cs]: Could use the `ISkinnableTarget` interface instead of the container implementation itself.

File: osu.Game/Skinning/SkinnableTargetComponent.cs (hunk 9 above)
Comment (peppy) [reply osu.Game/Skinning/SkinnableTargetComponent.cs:6]: I've applied a bit of renaming, see how things feel now.

**Review by frenzibyte — COMMENTED** [review]
Few small things I've noticed as well while I was checking on this, otherwise, this looks good for me as initial support.

File: osu.Game/Skinning/SkinnableTargetWrapper.cs [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+        {
+            RelativeSizeAxes = Axes.Both;
+        }
+
+        protected override void LoadComplete()
+        {
+            base.LoadComplete();
+
+            // schedule is required to allow children to run their LoadComplete and take on their correct sizes.
+            Schedule(() => applyDefaults?.Invoke(this));
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game/Skinning/SkinnableTargetWrapper.cs:43]: Shouldn't make a difference but I'll change it to read better.

File: osu.Game/Skinning/SkinnableTargetWrapper.cs (hunk 10 above)
Comment (smoogipoo) [reply osu.Game/Skinning/SkinnableTargetWrapper.cs:14]: How about

```suggestion
    /// <summary>
    /// A container which groups the elements of a <see cref="SkinnableTargetContainer"/> into a single object.
    /// Optionally also applies a default layout to the elements.
    /// </summary>
```

File: osu.Game/Skinning/SkinnableTargetComponent.cs (hunk 9 above)
Comment (smoogipoo) [reply osu.Game/Skinning/SkinnableTargetComponent.cs:6]: Much better. There's still a bit of weirdness regarding `SkinnableDrawable` that I forgot was a thing, but it's more digestible now. Also `SkinnableTargetWrapper` is a bit weird now that there's a `SkinnableTargetContainer` - perhaps a better name is `SkinnableTargetElementContainer` or something like that.

Comment (peppy) [reply osu.Game/Skinning/SkinnableTargetComponent.cs:6]: `SkinnableDrawable` should mostly be out of the equation with the new classes. It's still something we can keep around into the future, though.

**Review by smoogipoo — DISMISSED** [review]

**Review by smoogipoo — APPROVED** [review]

**Review by frenzibyte — APPROVED** [review]
